AstarteDeviceSDKCSharp/Data/AstarteFailedMessageStorage.cs

```diff
@@ -189,86 +196,91 @@ public bool IsExpired(long expire)
-                                 message.ApplicationMessage.Payload,
-                                 (int)message.ApplicationMessage.QualityOfServiceLevel,
-                                 message.Id
-                             ).Wait();
-                        }
-                    }
-                    else if (retention.Value == "VOLATILE")
+                if (retention.Value == "STORED")
+                {
+                    if (message.ApplicationMessage.MessageExpiryInterval > 0)
                     {
-
-                        if (message.ApplicationMessage.MessageExpiryInterval > 0)
-                        {
-                            InsertVolatile
+                        await InsertStored
                             (
                                 message.ApplicationMessage.Topic,
                                 message.ApplicationMessage.Payload,
                                 (int)message.ApplicationMessage.QualityOfServiceLevel,
                                 message.Id,
                                 (int)message.ApplicationMessage.MessageExpiryInterval
                             );
-                        }
-                        else
-                        {
-                            InsertVolatile
+                    }
+                    else
+                    {
+                        await InsertStored
                             (
                                 message.ApplicationMessage.Topic,
                                 message.ApplicationMessage.Payload,
                                 (int)message.ApplicationMessage.QualityOfServiceLevel,
                                 message.Id
                             );
-                        }
                     }
-                    stored.Add(message);
+                }
+                else if (retention.Value == "VOLATILE")
+                {
 
+                    if (message.ApplicationMessage.MessageExpiryInterval > 0)
+                    {
+                        InsertVolatile
+                        (
+                            message.ApplicationMessage.Topic,
+                            message.ApplicationMessage.Payload,
+                            (int)message.ApplicationMessage.QualityOfServiceLevel,
+                            message.Id,
+                            (int)message.ApplicationMessage.MessageExpiryInterval
+                        );
+                    }
+                    else
+                    {
+                        InsertVolatile
+                        (
+                            message.ApplicationMessage.Topic,
+                            message.ApplicationMessage.Payload,
+                            (int)message.ApplicationMessage.QualityOfServiceLevel,
+                            message.Id
+                        );
+                    }
                 }
+
+                lock (_storeLock)
+                {
+                    stored.Add(message);
```

Comment:
What do you think to use a Concurrent collection? it is thread-safe collection class that allows multiple threads accessing the collection concurrently.
[System.Collections.Concurrent](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent?view=net-6.0)

Reply:
I like your suggestion on PR. But, after additional testing we found that latency could be reason why streaming data is slower in some cases. 
First, I found that messages list named store can be removed now, as we use it to avoid EF exceptions. Since we don't use EF more, I deleted it.